rueidisprob/bloomfilter.go

```diff
@@ -0,0 +1,347 @@
+func NewBloomFilter(
+	client rueidis.Client,
+	name string,
+	expectedNumberOfItems uint,
+	falsePositiveRate float64,
+) (BloomFilter, error) {
+	if len(name) == 0 {
+		return nil, ErrEmptyName
+	}
+
+	if falsePositiveRate <= 0 {
+		return nil, ErrFalsePositiveRateLessThanEqualZero
+	}
+	if falsePositiveRate > 1 {
+		return nil, ErrFalsePositiveRateGreaterThanOne
+	}
+
+	size := numberOfBits(expectedNumberOfItems, falsePositiveRate)
+	if size == 0 {
+		return nil, ErrBitsSizeZero
+	}
+	if size > maxSize {
+		return nil, ErrBitsSizeTooLarge
+	}
+	hashIterations := numberOfHashFunctions(size, expectedNumberOfItems)
+
+	// NOTE: https://redis.io/docs/reference/cluster-spec/#hash-tags
+	bfName := "{" + name + "}"
+	counterName := bfName + ":c"
+	return &bloomFilter{
+		client:              client,
+		name:                bfName,
+		counter:             counterName,
+		hashIterations:      hashIterations,
+		hashIterationString: strconv.FormatUint(uint64(hashIterations), 10),
+		size:                size,
+		addMultiScript:      rueidis.NewLuaScript(addMultiScript),
+		addMultiKeys:        []string{bfName, counterName},
+		existsMultiScript:   rueidis.NewLuaScript(existsMultiScript),
+		existsMultiKeys:     []string{bfName},
+	}, nil
+}
+
+func numberOfBits(n uint, r float64) uint {
+	return uint(math.Ceil(-float64(n) * math.Log(r) / math.Pow(math.Log(2), 2)))
+}
+
+func numberOfHashFunctions(s uint, n uint) uint {
+	return uint(math.Round(float64(s) / float64(n) * math.Log(2)))
+}
+
+func (c *bloomFilter) Add(ctx context.Context, key string) error {
+	return c.AddMulti(ctx, []string{key})
+}
+
+func (c *bloomFilter) AddMulti(ctx context.Context, keys []string) error {
+	if len(keys) == 0 {
+		return nil
+	}
+
+	indexes := c.indexes(keys)
+
+	var args []string
+	args = append(args, c.hashIterationString)
+	args = append(args, indexes...)
+
+	resp := c.addMultiScript.Exec(ctx, c.client, c.addMultiKeys, args)
+	if resp.Error() != nil {
+		return resp.Error()
+	}
+
+	return nil
+}
+
+func (c *bloomFilter) indexes(keys []string) []string {
+	var allIndexes []string
+	for _, key := range keys {
+		indices := indexes([]byte(key), c.hashIterations, c.size)
+		for _, index := range indices {
+			allIndexes = append(allIndexes, strconv.FormatUint(index, 10))
+		}
+	}
+	return allIndexes
+}
+
+func (c *bloomFilter) Exists(ctx context.Context, key string) (bool, error) {
+	exists, err := c.ExistsMulti(ctx, []string{key})
+	if err != nil {
+		return false, err
+	}
+
+	return exists[0], nil
+}
+
+func (c *bloomFilter) ExistsMulti(ctx context.Context, keys []string) ([]bool, error) {
+	if len(keys) == 0 {
+		return nil, nil
+	}
+
+	indexes := c.indexes(keys)
+
+	var args []string
+	args = append(args, c.hashIterationString)
+	args = append(args, indexes...)
+
+	resp := c.existsMultiScript.Exec(ctx, c.client, c.existsMultiKeys, args)
+	if resp.Error() != nil {
+		return nil, resp.Error()
+	}
+
+	is, err := resp.AsIntSlice()
```

Comment:
```suggestion
	is, err := resp.ToArray()
```

`AsIntSlice` will allocate a new slice which is costly if it is a temporary storage to a bool slice. Unfortunately, we don't have `AsBoolSlice` yet. We can instead use `ToArray` to avoid allocation.

Reply:
[d037dad](https://github.com/redis/rueidis/pull/493/commits/d037dadc6557908a437bdca34b87112d59ac6d1d)

Oh, Thank you!